codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -445,8 +506,9 @@ public FlowControlledBase(final ChannelHandlerContext ctx, final Http2Stream str
             this.endOfStream = endOfStream;
             this.stream = stream;
             this.promise = promise;
-            // Ensure error() gets called in case something goes wrong after the frame is passed to Netty.
-            promise.addListener(this);
+            if (!promise.isVoid()) {
```

Comment:
That's a good point. To ensure that `error` is called regardless of who fails/completes the promise, setting the listener in the base class (as is done now) is probably better.